### DISCUSSION TITLE
On the problem of linearization

### DISCUSSION THREAD

**Discussion by hzzz123** [OpenFAST/openfast#2112] [Q&A]
I linearized with Openfast version 3.5, and successfully generated 36. lin files.  But the wind speed in these files shows ********m/s instead of numerical value.What went wrong?
Best regards,

**Comment by jjonkman** [comment]
Dear @hzzz123,

A numerical value will show up as stars in the output file if the value is outside the range of possible values as defined by the format statement for that output.  Can you clarify what wind speed you are linearizing about?  What have you set `OutFmt` to in the OpenFAST primary (._fst_) input file?

Best regards,

**Reply by hzzz123** [reply]
Dear @jjonkman ,
My wind speed setting is a steady wind speed of 18m/s. OutFmt is the default setting "ES15.7E2" and has not been changed. Is there anything wrong with my settings?
Best regards,

**Reply by jjonkman** [reply]
Dear @hzzz123,

Thanks for clarifying, but I'm not sure what the issue is.

Are other stars showing up in the linearization output file(s)?  Are the linearized matrices making sense?  Do you have this issue every time you linearize or only in specific cases?  Which version of OpenFAST are you using?

Best regards,

**Reply by hzzz123** [reply]
Dear @jjonkman ,
In the linearization files, only the wind speed appears star.  When I changed star to 18.0000, I can successfully use fx_mbc3 function.
I use version 3.5 of Openfast to linearize 5MW _ Itibarge _ DLL _ W TURB _ Wavesirr.
This is my first linearization, and there are still a few questions. 1. The wind speed of 18m/s is linearized, so is the linearized model inaccurate at other wind speeds? 2. Is there any documentation for linearization setting for reference?
Best regards,

**Reply by hzzz123** [reply]
Dear @jjonkman ,
After I use fx_mbc3 function, "AD User property on blade 1, node 2,-","AD User property on blade 2, node 2,-"and "AD User property on blade 3, node 2,-"correspond to columns in matrix B and matrix D have values. Is this normal? Even if they will be deleted after the model is reduced.
Best regards,

**Reply by jjonkman** [reply]
Dear @hzzz123,

Perhaps the starts in the wind speed are caused by bugs in AeroDyn or InflowWind that were fixed in v3.5.1, 3.5.2 or the soon-to-be-released v3.5.3?

In general, a linearized model generated from OpenFAST is valid for small perturbations about the operating point about which the model was linearized.  What constitutes a "small perturbation" depends on the degree of nonlinearity in the underlying nonlinear model of OpenFAST.  For wind speed, I would say it common to linearize at different wind speeds, e.g., in 2 m/s increments.

Regarding references on linearization, what application of linearization are you pursuing, e.g., modal analysis, generation of Campbell diagrams, stability analysis, controls design? 

Best regards,

**Reply by hzzz123** [reply]
Dear @jjonkman ,
I may have found the reason. When I used Openfast.exe versions 3.4.1 and 3.5.1, the wind speed in the generated linearization file was displayed normally, so it was a version BUG.
I use linearization for control design, so what references are there?
I still have a question. When linearized at the wind speed of 18m/s, should the value of "VS_RtGnSp" be 999e-9 or 1173.7? This seems to affect the linearization result.
Best regards,

**Reply by jjonkman** [reply]
Dear @hzzz123,

For linearization applied to controls design, I would say that Alan Wright's Ph.D. Thesis-turned NREL report is likely of interest to you: https://www.nrel.gov/docs/fy04osti/35816.pdf (although the version of FAST used in this report is a bit outdated now).

For more modern controls references, I would review ROSCO: https://github.com/NREL/ROSCO.

When linearizing OpenFAST in region 3 (above rated but before cut-out), we generally suggest linearizing with constant torque, so, I would set `VS_RtGnSp` = 999e-9 in that case.

Best regards,

**Reply by hzzz123** [reply]
Dear @jjonkman ,
Thank you for providing references. Regarding the setting of VS_RtGnSp= 999e-9, the following error occurred. Why is this?
In addition, how much should BlPitch(i) be set for the linearization setting of Area 3 at the wind speed of 18m/s?
Best regards,
```
 Time: 60 of 200 seconds.  Estimated final completion at 22:47:27.
The BEM solution is being turned off due to low TSR.  (TSR = 1.9994). This warning will not be
 Time: 70 of 200 seconds.  Estimated final completion at 22:46:55.

FAST_Solution:FAST_AdvanceStates:SrvD_UpdateStates:CalculateTorque:u%HSS_Spd is negative. Simple
variable-speed control model is not valid for motoring situations.

 OpenFAST encountered an error at simulation time 72.597 of 200 seconds.
 Simulation error level: FATAL ERROR
```

**Reply by jjonkman** [reply]
Dear @hzzz123,

This error implies to me that your simulation is not tending toward a steady state condition.  Have you enabled `CalcSteady` = TRUE with `TrimCase` = 3 for this region 3 linearization?

Best regards,

**Reply by hzzz123** [reply]
Dear @jjonkman ,
Yes, of course. 
And when VS_RtGnSp= 1173.7, it can be linearized.
Best regards,

**Reply by jjonkman** [reply]
Dear @hzzz123,

Can I presume you are simulating the NREL 5-MW baseline turbine and have set `VS_RtTq` = 43093.55 Nm (i.e., the rated torque) and that your target steady-state rotor speed is 12.1 rpm?  If you plot the time series of rotor speed, blade pitch, and generator torque during the steady-state calculation, does it appear that the solution is converging?

Best regards,

**Reply by jjonkman** [reply]
I see that the issues with the wind speed showing up as stars (***) was discussed in https://github.com/OpenFAST/openfast/issues/1622 and fixed in https://github.com/OpenFAST/openfast/pull/1655, which was released with OpenFAST 3.5.1.
